howso/residuals.amlg

```diff
@@ -118,73 +118,270 @@
+												(compute_on_contained_entities (list
+													(query_value_masses current_feature (null) (contains_index !numericNominalFeaturesMap current_feature) )
+												))
+										))
 									)
+									;do not max cap nominal deviations if less than 2 classes have been trained
+									(if (<= (size class_counts_map) 1)
+										(max
+											(call !GetSharedDeviationsGroupValues (assoc
+												aggregation "max"
+												feature_group_to_retrieve current_feature
+												group_value_location (call !cachedFeatureMinResidualMap)
+											))
+											(current_value)
+											(call !GetSharedDeviationsGroupValues (assoc
+												aggregation "max"
+												feature_group_to_retrieve current_feature
+												group_value_location (call !userSpecifiedFeatureErrorsMap)
+											))
+										)
 
-									(max
-										(get !cachedFeatureMinResidualMap (current_index))
-										(current_value)
-										(get !userSpecifiedFeatureErrorsMap (current_index))
+										;else cap the max nominal deviation
+										(let
+											(assoc total_count (apply "+" (values class_counts_map)) )
+											(min
+												;nominal max deviation is the sum of: each class's probability multiplied by probability of getting it wrong
+												(apply "+"
+													(map
+														(lambda (let
+															(assoc class_prob (* (/ (current_value 1) total_count)) )
+															(* class_prob (- 1 class_prob))
+														))
+														(values class_counts_map)
+													)
+												)
+
+												(max
+													(call !GetSharedDeviationsGroupValues (assoc
+														aggregation "max"
+														feature_group_to_retrieve current_feature
+														group_value_location (call !cachedFeatureMinResidualMap)
+													))
+													(current_value)
+													(call !GetSharedDeviationsGroupValues (assoc
+														aggregation "max"
+														feature_group_to_retrieve current_feature
+														group_value_location (call !userSpecifiedFeatureErrorsMap)
+													))
+												)
+											)
+										)
 									)
 								)
+								(max
+									(call !GetSharedDeviationsGroupValues (assoc
+										aggregation "max"
+										feature_group_to_retrieve (current_index 1)
+										group_value_location (call !cachedFeatureMinResidualMap)
+									))
+									(call !GetSharedDeviationsGroupValues (assoc
+										aggregation "max"
+										feature_group_to_retrieve (current_index 1)
+										group_value_location (call !userSpecifiedFeatureErrorsMap)
+									))
+									(current_value)
+								)
 							)
 						)
+						(call !ReduceSharedDeviationsMapping (assoc map_to_reduce (get feature_residuals_map "residual_map") ))
 					)
+			)
+
+			(assign (assoc
+				temp_residual_map
+					(call !ExpandForSharedDeviations (assoc
+						compressed_values temp_residual_map
+					))
+			))
+
+			temp_residual_map
```

Comment:
if we assign to a variable and then immediately return it, (as long as there are no other things being statefully changed), we can instead just return the value that we are assigning with immediately.

so instead of:
``` 
(assign (assoc temp_residual_map (call ....)  ))
temp_residual_map
``` 

you should probably just do
```
(call ...)
```